networkit/cpp/embedding/LearnEmbedding.cpp

```diff
@@ -0,0 +1,256 @@
+            // negative sampling
+            for (index j = 0; j < negSamN + 1; ++j) {
+                node target;
+                int label;
+                if (j == 0) {
+                    target = word;
+                    label = 1;
+                } else {
+                    target = as.sample();
+                    if (target == word) {
+                        continue;
+                    }
+                    label = 0;
+                }
+                double product = 0;
+                for (index i = 0; i < dimensions; ++i) {
+                    product += synPos[currWord][i] * synNeg[target][i];
+                }
+                double grad; // Gradient multiplied by learning rate
+                if (product > maxExp) {
+                    grad = (label - 1) * alpha;
+                } else if (product < -maxExp) {
+                    grad = label * alpha;
+                } else {
+                    double exp = std::exp(product);
+                    grad = (label - 1 + 1 / (1 + exp)) * alpha;
+                }
+
+                for (index i = 0; i < dimensions; ++i) {
+                    eV[i] += grad * synNeg[target][i];
+#ifndef NETWORKIT_OMP2
+#pragma omp atomic write
+#endif // NETWORKIT_OMP2
+                    synNeg[target][i] = synNeg[target][i] + grad * synPos[currWord][i];
+                }
+            }
+
+            for (index i = 0; i < dimensions; ++i) {
+#ifndef NETWORKIT_OMP2
+#pragma omp atomic write
+#endif // NETWORKIT_OMP2
+                synPos[currWord][i] = synPos[currWord][i] + eV[i];
+            }
+        }
+#ifndef NETWORKIT_OMP2
+#pragma omp atomic write
+#endif // NETWORKIT_OMP2
+        wordCntAll = wordCntAll + 1;
+    }
+}
+
+Embeddings learnEmbeddings(AllWalks &allWalks, count nn, count dimensions, count winSize,
+                           count iterations) {
+    count walks = allWalks.size();           // number of walks
+    count nodesPerWalk = allWalks[0].size(); // number of nodes per walk
+
+    std::vector<node> renumbered(nn, none);
+    std::vector<node> original(nn, none);
+    count nodes = 0;
+    for (index i = 0; i < walks; ++i) {
+        for (index j = 0; j < nodesPerWalk; ++j) {
+            node &nd = allWalks[i][j];
+            if (renumbered[nd] != none) {
+                nd = renumbered[nd];
+            } else {
+                renumbered[nd] = nodes;
+                original[nodes] = nd;
+                nd = nodes++;
+            }
+        }
+    }
+
+    std::vector<count> vocab = learnVocab(allWalks, nodes);
+
+    Embeddings embeddings(nodes, std::vector<float>(dimensions));
+
+    Embeddings synNeg = initNegEmb(vocab, dimensions);
+    Embeddings synPos = initPosEmb(vocab, dimensions);
+    AliasSampler as = vocabSampler(vocab);
+
+    double alpha = startAlpha; // learning rate
+    count wordCntAll = 0;
+
+    ModelData model(allWalks, dimensions, winSize, iterations, as, wordCntAll, alpha, synNeg,
+                    synPos);
```

Comment:
`wordCntAll` seems to be unused afterwards, why is this passed to the model as reference? Same for `alpha`.

Reply:
`alpha` is reduced along the progress of `wordCntAll`.